abo-abo/swiper
# [WIP] Interactively specify switches

Closes #1408 

## Purpose
Enable users to interactively specify extra switches along with the search pattern when invoking `counsel-ag`.

For example, `M-x counsel-ag`:

    [~/.emacs.d] ag: --elisp --ignore site-lisp -- let

## UI
Assume `counsel-ag-command` is set to `ag --nocolor --nogroup`.

- It behaves the same as before when the query starts with anything but a dash:
  `M-x counsel-ag` `abc` ↵
  Produces `ag --nocolor --nogroup abc`

- Regex are escaped:
  `M-x counsel-ag` `abc[^d].` ↵
  Produces `ag --nocolor --nogroup abc\[\^d\].`

- Starting with `-` or `--` passes the text as switches until a ` -- `:
  `M-x counsel-ag` `--elisp -- abc` ↵
  Produces `ag --nocolor --nogroup --elisp abc`

- Initial dashes need to be escaped if they are not switches:
  `M-x counsel-ag` `\-abc` ↵
  Produces `ag --nocolor --nogroup \\-abc`

- Switches are not escaped. Queries are:
  `M-x counsel-ag` `--elisp -- abc[^d].` ↵
  Produces `ag --nocolor --nogroup  --elisp abc\[\^d\].`

- Using a prefix argument behaves the same as before:
  `C-u M-x counsel-ag` `ag --nocolor --nogroup --elisp` ↵
  `abc` ↵
  Produces `ag --nocolor --nogroup --elisp abc`

## Commentary
Except for the two bugs below it seems to work OK, but more testing is welcome.
I'm looking for early feedback on the code and functionality and any bugs that I might have missed.

~~I don't have `rip-grep` so I don't know how it interacts with `counsel-rg`.~~ It looks like it works fine with `counsel-rg`.

## How to test
1. Clone the [repo](https://github.com/DEADB17/swiper).
2. Check out [`dash-dash-opts`](https://github.com/DEADB17/swiper/tree/dash-dash-opts).
3. Open [`ivy.el`](https://github.com/DEADB17/swiper/blob/dash-dash-opts/ivy.el) and `M-x emacs-lisp-byte-compile-and-load`.
4. Open [`counsel.el`](https://github.com/DEADB17/swiper/blob/dash-dash-opts/counsel.el) and `M-x emacs-lisp-byte-compile-and-load`.
5. Call ag: `M-x counsel-ag`.

## Known bugs
- [x] Fixed: ~~`ivy-occur` is broken.~~
- [x] Fixed: ~~The minimum character count in the `ag-prompt` is negative when switches are used.~~

## Contribution checklist
- [x] `make compile`
- [x] `make checkdoc`

## Maybe
- [ ] Add tests for the new functionality
- [ ] Add user documentation for the new functionality
